lightstep/sdk/metric/internal/viewstate/base_instrument.go

```diff
@@ -96,22 +96,40 @@ func (metric *instrumentBase[N, Storage, Auxiliary, Methods]) mergeDescription(d
 	}
 }
 
-func attributeFilter(viewf *attribute.Filter) attribute.Filter {
-	return func(kv attribute.KeyValue) bool {
-		if kv.Key == "" {
+// isValidAttribute supports filtering invalid attributes.  Note, this
+// should be fast, trye not to allocate!  TODO: match the
+// specification.
+func isValidAttribute(kv attribute.KeyValue) bool {
+	return kv.Key != ""
+}
+
+func (metric *instrumentBase[N, Storage, Auxiliary, Methods]) invalidAttributeFilter(kv attribute.KeyValue) bool {
+	return isValidAttribute(kv) && (metric.keysFilter == nil || (*metric.keysFilter)(kv))
+}
+
+func (metric *instrumentBase[N, Storage, Auxiliary, Methods]) applyKeysFilter(kvs attribute.Set) attribute.Set {
+	invalidFilter := false
+	for iter := kvs.Iter(); iter.Next(); {
+		kv := iter.Attribute()
+		if !isValidAttribute(kv) {
 			doevery.TimePeriod(time.Minute, func() {
 				otel.Handle(fmt.Errorf("use of empty attribute key, e.g., with value %q", kv.Value.Emit()))
 			})
-			return false
+			invalidFilter = true
+			break
 		}
-
-		return viewf == nil || (*viewf)(kv)
 	}
-}
 
-func (metric *instrumentBase[N, Storage, Auxiliary, Methods]) applyKeysFilter(kvs attribute.Set) attribute.Set {
-	kvs, _ = attribute.NewSetWithFiltered(kvs.ToSlice(), attributeFilter(metric.keysFilter))
-	return kvs
+	if !invalidFilter && metric.keysFilter == nil {
+		return kvs
+	}
+	var res attribute.Set
+	if !invalidFilter {
+		res, _ = kvs.Filter(*metric.keysFilter)
+	} else {
+		res, _ = kvs.Filter(metric.invalidAttributeFilter)
```

Comment:
Is there anything we can do here in the case of an invalid attribute to inform the user?

Reply:
Yes, already done on line 116.